$\large{\textbf{Exercise 4:}}$

$\min \ -z_1 + z_2 - z_3 - 4z_4 \\ \text{ s.t. } z_1\geq -5, z_2 \geq 2, \\  z_1 +z_2-z_3 + z_4 \leq 8 \\ -z_1 - z_2 + z_3 - 2 z_4 \leq 100, \\ z_1+z_2 + 2z_3+z_4 \leq -11$.  

In [1]:
!pip install -q pyomo

     |████████████████████████████████| 9.6 MB 6.4 MB/s 
     |████████████████████████████████| 49 kB 5.4 MB/s 


In [2]:
from pyomo.environ import *

In [3]:
#create a model
model = ConcreteModel()

In [4]:
#declare variable
model.z1 = Var(domain = Reals)
model.z2= Var(domain = Reals)
model.z3 = Var(domain = Reals)
model.z4 = Var(domain = Reals)


In [5]:
#declare objective
model.cost = Objective( expr = -1*model.z1 + 1*model.z2 - 1*model.z3 - 4*model.z4,sense = minimize)

In [6]:
#declare constraints
model.const1 = Constraint(expr = 1*model.z1 >= -5)
model.const2 = Constraint(expr = model.z2 >= 2)
model.const3 = Constraint(expr =model.z1 + model.z2 - model.z3 + model.z4 <= 8)
model.const4 = Constraint(expr = -1*model.z1 - 1*model.z2 + 1*model.z3 -2*model.z4 <= 100)
model.const5 = Constraint(expr = 1*model.z1 + 1*model.z2 +2*model.z3 + 1*model.z4 <= -11)



In [7]:
!apt-get install -y -qq glpk.utils

Selecting previously unselected package libsuitesparseconfig5:amd64.
(Reading database ... 155673 files and directories currently installed.)
Preparing to unpack .../libsuitesparseconfig5_1%3a5.1.2-2_amd64.deb ...
Unpacking libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libamd2:amd64.
Preparing to unpack .../libamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libcolamd2:amd64.
Preparing to unpack .../libcolamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libcolamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libglpk40:amd64.
Preparing to unpack .../libglpk40_4.65-1_amd64.deb ...
Unpacking libglpk40:amd64 (4.65-1) ...
Selecting previously unselected package glpk-utils.
Preparing to unpack .../glpk-utils_4.65-1_amd64.deb ...
Unpacking glpk-utils (4.65-1) ...
Setting up libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Setting up libcolamd2:amd64 (1:5.1.2-2) ...
Setting up libamd2:amd64 

In [8]:
model.pprint()

4 Var Declarations
    z1 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    z2 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    z3 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals
    z4 : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :  None :  None :  None : False :  True :  Reals

1 Objective Declarations
    cost : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : minimize : - z1 + z2 - z3 - 4*z4

5 Constraint Declarations
    const1 : Size=1, Index=None, Active=True
        Key  : Lower : Body : Upper : Active
        None :  -5.0 :   z1 :  +Inf :   True
    const2 : Size=1, Index=None, Active

In [9]:
SolverFactory('glpk',excutable = '/usr/bin/glpsol').solve(model).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: -5.33333333333333
  Upper bound: -5.33333333333333
  Number of objectives: 1
  Number of constraints: 6
  Number of variables: 5
  Number of nonzeros: 15
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.014966487884521484
# ----------------------------------------------------------
#   Solution Information
# -----------

In [10]:
print('\ncost = ', model.cost())
print('\nDecision Variables')
print('z1 = ', model.z1())
print('z2 = ', model.z2())
print('z3 = ', model.z3())
print('z4 = ', model.z4())
print('\nconstraints')
print('consraint 1 = ', model.const1())
print('consraint 2 = ', model.const2())
print('consraint 3 = ', model.const3())
print('consraint 4 = ', model.const4())
print('consraint 5 = ', model.const5())



cost =  -5.333333333333348

Decision Variables
z1 =  -5.0
z2 =  2.0
z3 =  -6.33333333333333
z4 =  4.66666666666667

constraints
consraint 1 =  -5.0
consraint 2 =  2.0
consraint 3 =  8.0
consraint 4 =  -12.66666666666667
consraint 5 =  -10.999999999999991


$\large{\textbf{observation:}}$

minimum value of objective is -5.33333333333
we get this value at 

  z1 = -5.0 

  z2 =2.0  

  z3 = -6.3333333   and 
  
  z4 = 4.66666666666